File: apps/forja/lib/features/iptv/iptv/data/iptv_portal_share.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:rust/rust.dart';

import 'models.dart';

/// Creates and resolves IPTV portal share tokens.
///
/// Self-contained `F1.` tokens (AES in Rust) — no pastebin / no server.
class IptvPortalShare {
  IptvPortalShare._();

  static const embeddedPrefix = 'F1.';

  static bool isEmbeddedToken(String raw) =>
      raw.trim().startsWith(embeddedPrefix);

  static bool isValidCode(String raw) =>
      isEmbeddedToken(raw) &&
      raw.trim().length > embeddedPrefix.length + 16;

  /// Encrypt portal credentials into a self-contained `F1.` token (Rust).
  static Future<String> createShare(IptvPortal portal) async {
    final token = RustLib.instance.iptvPortalShareEncode(
      portal.url,
      portal.username,
      portal.password,
    );
    if (token.isEmpty || !isEmbeddedToken(token)) {
      throw StateError('Could not create share code');
    }
    return token;
  }

  /// Resolve an embedded `F1.` token.
  static Future<IptvPortal?> resolveShare(String rawCode) async {
    final trimmed = rawCode.trim();
    if (!isEmbeddedToken(trimmed)) return null;
    return _resolveEmbedded(trimmed);
  }

  static IptvPortal? _resolveEmbedded(String token) {
    try {
      final jsonStr = RustLib.instance.iptvPortalShareDecode(token);
      if (jsonStr.isEmpty) return null;
      final decoded = jsonDecode(jsonStr);
      if (decoded is! Map) return null;
      final map = Map<String, dynamic>.from(decoded);
      final url = (map['url'] as String?)?.trim() ?? '';
      final username = (map['username'] as String?)?.trim() ?? '';
      final password = (map['password'] as String?)?.trim() ?? '';
      if (url.isEmpty || username.isEmpty || password.isEmpty) return null;
      return IptvPortal(
        url: url,
        username: username,
        password: password,
        source: 'Shared',
      );
    } catch (e, st) {
      debugPrint('[IptvPortalShare] decode failed: $e\n$st');
      return null;
    }
  }
}

File: apps/admin/src/lib/iptv-portal-share.ts
/** Client share-code helpers (same crypto as apps/web). */

const EMBEDDED_PREFIX = 'F1.'
const KEY_MATERIAL = 'forja-iptv-share-embedded-v1'

export type SharePortal = {
  url: string
  username: string
  password: string
}

export function isEmbeddedShareToken(raw: string): boolean {
  return raw.trim().startsWith(EMBEDDED_PREFIX)
}

export function formatShareCode(raw: string): string {
  return raw.trim()
}

async function sha256(text: string): Promise<Uint8Array> {
  const data = new TextEncoder().encode(text)
  const digest = await crypto.subtle.digest('SHA-256', data)
  return new Uint8Array(digest)
}

function bytesToBase64Url(bytes: Uint8Array): string {
  let binary = ''
  for (const byte of bytes) binary += String.fromCharCode(byte)
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '')
}

function base64UrlToBytes(value: string): Uint8Array {
  const clean = value.replace(/\n/g, '').trim()
  const padded = clean.replace(/-/g, '+').replace(/_/g, '/')
  const padLen = (4 - (padded.length % 4)) % 4
  const binary = atob(padded + '='.repeat(padLen))
  const out = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i += 1) out[i] = binary.charCodeAt(i)
  return out
}

async function embeddedKey(): Promise<CryptoKey> {
  const raw = await sha256(KEY_MATERIAL)
  return crypto.subtle.importKey('raw', raw, { name: 'AES-CBC' }, false, [
    'encrypt',
    'decrypt',
  ])
}

async function encodeEmbeddedShare(portal: SharePortal): Promise<string> {
  const url = portal.url.trim()
  const username = portal.username.trim()
  const password = portal.password.trim()
  if (!url || !username || !password) {
    throw new Error('url, username, and password are required')
  }
  const plain = new TextEncoder().encode(
    JSON.stringify({
      v: 1,
      url,
      username,
      password,
    }),
  )
  const iv = crypto.getRandomValues(new Uint8Array(16))
  const key = await embeddedKey()
  const cipher = await crypto.subtle.encrypt({ name: 'AES-CBC', iv }, key, plain)
  const packed = new Uint8Array(16 + cipher.byteLength)
  packed.set(iv, 0)
  packed.set(new Uint8Array(cipher), 16)
  return `${EMBEDDED_PREFIX}${bytesToBase64Url(packed)}`
}

async function decodeEmbeddedShare(
  token: string,
): Promise<SharePortal | null> {
  try {
    const trimmed = token.trim()
    if (!trimmed.startsWith(EMBEDDED_PREFIX)) return null
    const packed = base64UrlToBytes(trimmed.slice(EMBEDDED_PREFIX.length))
    if (packed.length < 32 || packed.length % 16 !== 0) return null
    const iv = packed.slice(0, 16)
    const cipherBytes = packed.slice(16)
    const key = await embeddedKey()
    const plainBuf = await crypto.subtle.decrypt(
      { name: 'AES-CBC', iv },
      key,
      cipherBytes,
    )
    const decoded = JSON.parse(new TextDecoder().decode(plainBuf)) as {
      url?: string
      username?: string
      password?: string
    }
    const url = decoded.url?.trim() ?? ''
    const username = decoded.username?.trim() ?? ''
    const password = decoded.password?.trim() ?? ''
    if (!url || !username || !password) return null
    return { url, username, password }
  } catch {
    return null
  }
}

/** Encrypt credentials into a self-contained `F1.` token (no server). */
export async function createPortalShare(portal: SharePortal): Promise<string> {
  return encodeEmbeddedShare(portal)
}

/** Resolve an `F1.` token. */
export async function resolvePortalShare(
  rawCode: string,
): Promise<SharePortal | null> {
  const trimmed = rawCode.trim()
  if (!isEmbeddedShareToken(trimmed)) return null
  return decodeEmbeddedShare(trimmed)
}
